2.7: New password-based authenticated encryption sample program pbcrypt

gilles-peskine-arm opened the pull request:
This PR introduces `programs/cipher/pbcrypt`, a new sample program to demonstrate password-based authenticated encryption.

This is meant to illustrate good practice, so please review accordingly. A few caveats:

* The program works in a specific way: streaming AEAD. Since the library doesn't support streaming CCM, the program doesn't support it.
* The use of PKCS#12 key derivation is hard-coded because the library doesn't have a generic interface to password-based key derivation.
* The program should build in all configurations for which the library builds, but it may not run usefully in some “exotic” configurations (e.g. GCM enabled but no block cipher that supports GCM).
* The program relies on argv and stdio. It does password-based key derivation so it isn't meant for microcontrollers anyway.

Provide a demo usage script and run it from `all.sh`.

Remove the badly designed `aescrypt2`. Fix #1906.

This PR is for 2.7. I'll forward-port it to 2.16+ once it's approved. I think this will mean the following changes;
* Add ChachaPoly to the `#ifdef` about `main` and to `cipher_is_aead`.
* Replace `mbedtls_zeroize` by `mbedtls_platform_zeroize`.
* Even in development, this is a sample program for the mbedtls API. In the PSA API, there's no password-based key derivation yet and multi-part AEAD is not implemented yet, but `key_ladder_demo` demonstrates symmetric key derivation and single-part AEAD.

mpg commented:
@gilles-peskine-arm  There appears to be a significant intersection with #2698 - adding sample usage scripts for programs and testing them in the CI. Does that mean the two PRs conflicts with each other? If that's the case, I'd suggest giving priority to #2698 and labeling this one as "needs: preceding PR".

gilles-peskine-arm commented:
The intersection was only adding something to `all.sh`. I've removed that part: #2698 is a better place for it since that's its main job. Now the PRs are fully independent, except insofar as the format of `pbcrypt_demo.sh` only makes full sense in combination with #2698.

mpg commented:
Ok, thanks!

gilles-peskine-arm commented on programs/cipher/pbcrypt_demo.sh:
```diff
@@ -0,0 +1,87 @@
+#!/bin/sh
+set -e -u
```
When #2698 is ready and backported, include `demo_common.sh` and adapt the code here accordingly.

mpg commented on programs/cipher/pbcrypt.c:
```diff
@@ -0,0 +1,720 @@
+        while ( 0 )
+
+/* Implementation that should never be optimized out by the compiler */
+static void mbedtls_zeroize( void *v, size_t n ) {
```
Note: in the meantime we introduced `mbedtls_platform_zeroize()`.

gilles-peskine-arm replied:
Not in 2.7, and already noted for the other branches (see PR description).

mpg replied:
Oh, right. Sorry for the noise then.

dgreen-arm commented on programs/cipher/pbcrypt.c:
```diff
@@ -0,0 +1,720 @@
+                                        (unsigned long long) header[7] ),
+                       payload_size <= SIZE_MAX,
+                       "Payload too large" );
+    metadata->payload_size = payload_size;
```
This line is causing failures on Windows due to the conversion from `unsigned long long` to `size_t`. Given the check above it should be fine, but the compiler is complaining about it.

gilles-peskine-arm commented:
Due to the imminent retirement of 2.7, I am closing this pull request. I have ported the program to 2.16 in https://github.com/ARMmbed/mbedtls/pull/4216. Compared to the 2.7 version, I added ChaChaPoly support and fixed a couple of minor bugs, all in new commits.
